### PR TITLE
ENH/DEV: Formula Viewer for Archive Viewer

### PR DESCRIPTION
**Pull request by aksharsarvesh**
Created class FormulaCurveItem. This accepts a formula (in the form of a string) and a dictionary of string : BasePlotCurveItem and populates a new data buffer with the function of the data being passed in. Can be used to graph functions of other functions, by passing FormulaCurveItems in the dictionary and referencing them in the formula string.

This is useful for the Archive Viewer because one of the requested functionalities is the ability to graph a function of one or more PVs rather than graphing only PVs themselves. As the new class extends BasePlotCurveItem, this is usable in other contexts where BasePlotCurveItems are being used.

### CODE REVIEW

**Comment by aksharsarvesh** [comment]
Hi there @YektaY @nstelter-slac @jbellister-slac I was wondering if one or more of you would be free to look at this when you get the chance? Hoping to finish this because I have a couple of other branches that are dependent on this. Thanks so much in advance!

**Comment by aksharsarvesh** [comment]
Hi @ZLLentz, yes sorry the documentation is sparse. I'll add to it soon, but I thought I'd respond first. 

> What's the expected usage here? Why is this implemented strictly for the archive viewer, rather than being implemented more broadly like the calc plugin is? 

One of the requested functionalities on the archive viewer was the ability to graph *functions* of PVs as opposed to just the PVs themselves. Currently, my project is focused on enhancing the archive viewer, so I made changes under that context.

I am happy to broaden that context to support usage outside of the archive viewer - would you have recommendations/suggestions for what you would like that to look like?. The FormulaCurveItem class extends a BasePlotCurveItem, which means tools that use baseplot will also be able to use the formulas. 

Currently we expect archived data, so one fix would be removing that expectation such that formulas could be used with timeplots as well. 

> What kinds of new/better/easier displays can we make with this PR?

If other projects would ever need the ability to graph functions of PVs, this class could be useful for them.

**Comment by nstelter-slac** [comment]
hi, thanks for adding the example and test!

+1 to updating the documentation,
so think it makes sense for me to wait for its update and then I can review shortly after

**Comment by aksharsarvesh** [comment]
So sorry, my supervisor is out today. I was able to find an archiver_timeplot.rst file in the repo that I could add the FormulaCurveItem to, but I'm not sure that is what you are looking for. Would you happen to know more specifically what you would want documentation-wise on this? Thanks in advance!

**Comment by nstelter-slac** [comment]
think for the docs its good to have general description of the new feature (something similar to what you added to this PR's description), and then explain generally how to use it (and be sure explain anything that could be confusing with its usage or arguments). 

could add a new section to archiver_timeplot.rst, or expand the 'Plot Usage' section perhaps

these are the docs I added for a new feature few months back, if you want a reference of what we've done in the past https://github.com/slaclab/pydm/pull/1052/files#diff-0619cbc99f71d1206dd6a1d53201fe0b0a30255027a50f6d99e771033d5c6497R86

**Comment by aksharsarvesh** [comment]
Ok, thank you very much. I think I am ready for review fully now

**Review by nstelter-slac — CHANGES_REQUESTED** [review]
nice patch! sry for the delay

added some suggestions I had, mainly I think there is some stuff that could make the main logic part a bit more readable and therefore maintainable.

i'd say we get should at least 1 (but ideally more) other reviewer on this before we can merge

thanks!

File: pydm/widgets/archiver_time_plot.py
Comment (aksharsarvesh) [reply pydm/widgets/archiver_time_plot.py:450]: Ok, check here if this is still an issue, thanks!

File: pydm/widgets/archiver_time_plot.py
Comment (aksharsarvesh) [reply pydm/widgets/archiver_time_plot.py]: I've added these changes; however, I have it set up on the Archive Viewer repo side that a formula is created iff the channel starts with f://, so such a check *shouldn't* be necessary unless somebody else is using the FormulaCurveItem object. Because the point of this is that potentially somebody else does, I think it makes sense regardless

File: pydm/widgets/archiver_time_plot.py
Comment (aksharsarvesh) [reply pydm/widgets/archiver_time_plot.py:424]: I believe I have now done this, please let me know if there's more

File: pydm/widgets/archiver_time_plot.py
Comment (nstelter-slac) [reply pydm/widgets/archiver_time_plot.py:497]: since it looks like we're doing the same logic for live and archive but just using different data, 
could we check if using live/archive and then set a common variable like 'pvData' to either live or archive set depending?

for example this chunk:
```
        for pv in self.pvs.keys():
            pv_archive_times = pvArchiveData[pv][0]
            pv_archive_values = pvArchiveData[pv][1]
            pv_current_index = pvIndices[pv]
            pvValues[pv] = pv_archive_values[pv_current_index]
            while pv_current_index < len(pv_archive_times) - 1 and pv_archive_times[pv_current_index] < self.minx:
                pvValues[pv] = pv_archive_values[pv_current_index]
                pv_current_index += 1
                # Shift starting indices for each row to our minimum
            pvIndices[pv] = pv_current_index
````
is the same as this chunk:
````
            for pv in self.pvs.keys():
                pv_live_times = pvLiveData[pv][0]
                pv_live_values = pvLiveData[pv][1]
                pv_current_index = pvIndices[pv]
                pvValues[pv] = pv_live_values[pv_current_index]
                while pv_current_index < len(pv_live_times) - 1 and pv_live_times[pv_current_index] < minx:
                    pvValues[pv] = pv_live_values[pv_current_index]
                    pv_current_index += 1

                pvIndices[pv] = pv_current_index
````


except for which map we are acting on. Could we just use a function arg to set some variable like 'pvData' to replace 'pvLiveData' and 'pvArchieveData'

File: examples/archiver_time_plot/formula_curve_example.py
Comment (aksharsarvesh) [reply examples/archiver_time_plot/formula_curve_example.py]: Haha thats funny I was just taking your word for it. Changed!

File: pydm/widgets/archiver_time_plot.py
Comment (aksharsarvesh) [reply pydm/widgets/archiver_time_plot.py]: Right, I understand. Done now. Thanks!

Comment (aksharsarvesh) [reply pydm/widgets/archiver_time_plot.py]: Right, now I understand. Changed, thanks!

File: examples/archiver_time_plot/formula_curve_example.py
Comment (aksharsarvesh) [reply examples/archiver_time_plot/formula_curve_example.py]: Haha my bad I was just taking your word for it.

**Review by nstelter-slac — APPROVED** [review]

File: pydm/widgets/archiver_time_plot.py
Comment (jbellister-slac) [inline pydm/widgets/archiver_time_plot.py:888]: I'm not sure I fully understand what the intent for this method is, but doesn't seem to be called anywhere. Maybe can be removed?

**Review by jbellister-slac — COMMENTED** [review]
Thanks for adding the additional context to the PR description, and thanks also for your patience while we take a look at this. It is looking very good!

My main thought on the first pass through is that there's quite a lot of data processing happening in a GUI widget here - I think most of it could be potentially moved to the data plugin layer.

One possible way of putting this together would be to use the calc plugin prefix. An end user can then enter in a formula like here, this example only uses a single PV but multiple PVs are fine too:

https://github.com/slaclab/pydm/blob/8b2742907fa9736663c09e890732d3ccb489c208/examples/calc/sin_cos_tan.ui#L60

The `archiver_plugin` would need to be modified to handle this case (using the code you've already written). Then the `archiver_time_plot` becomes simpler after this as it can just say give me the result of this calculation between these PVs and display it. It no longer has to maintain its own list/dict of curve items for its calculations, that's all handled for it now. The live data functionality is also free at that point since it's just already what the time plot does using the calc plugin, and maybe the new `FormulaCurveItem` can just inherit most of its functionality from `ArchivePlotCurveItem` at that point. I think this could be cleaner and easier to maintain.

That said, it might be easiest to get some of these PRs merged in before any potential refactor so the relevant code is all moved at once. And I might still need to get some more context to fully understand the precise use case that we're going for here since it looks like it involves the archiver application built on top of PyDM. But the code looks great so far and works well! This will be nice functionality to add.

**Comment by jbellister-slac** [comment]
Also if it would be helpful, I'm happy to drop by early next week to chat in person about any of this, or on zoom would be fine too.

File: pydm/widgets/timeplot.py
Comment (aksharsarvesh) [reply pydm/widgets/timeplot.py]: Agreed, and fixed!

File: pydm/widgets/archiver_time_plot.py
Comment (aksharsarvesh) [reply pydm/widgets/archiver_time_plot.py]: It was bad form - it was called from the ArchiveViewer side, but I have now removed it and everything functions as I want it to. Thank you for the note!

**Comment by aksharsarvesh** [comment]
Texting you on Slack about meeting next week for this! Thank you

**Review by YektaY — APPROVED** [review]
This looks good to me

**Pull request merged**
